### PR TITLE
feat: Allow reuse of variable definitions between policies

### PR DESCRIPTION
**Pull request by haines**
This PR introduces a new policy type that provides variable definitions for export:

```yaml
apiVersion: api.cerbos.dev/v1
exportVariables:
  name: my_common_variables
  definitions:
    foo: bar
```

The other policy types can then import those variable definitions:

```yaml
apiVersion: api.cerbos.dev/v1
resourcePolicy:
  variables:
    import:
      - my_common_variables
    local:
      baz: qux
  # ...
```

The top-level `variables` field is deprecated, and will be merged with the `local` definitions.

Duplicate variable definitions produce an error, just like derived roles.

### CODE REVIEW

**Comment by haines** [comment]
@charithe yep, that was my first instinct too 😄 Unfortunately `variables` is a top-level field on the `Policy` message, so we can't (re)use it in the `oneof policy_type` - we'll need to have a different name.

Also, if we have `definitions` in the variables policy, the `Policy` message still permits a top-level `variables` key, which seems potentially confusing.

We could add validation to prevent the use of `variables` in variables policies, if we went down the `definitions` route, though. We could even deprecate the top-level `variables` field and make `variables` and `importVariables` fields on the policy types that support them.

This brings me to another point - should variables policies support `importVariables` themselves (creating transitive dependencies)? Because the field is defined at the top level `Policy` message in order to live next to locally-defined `variables`, that would imply yes, but it does add some complexity.

**Comment by haines** [comment]
>(We should probably move this discussion to the issue to keep the context in one place.)

There isn't much on the issue at this point so I think the context is here... happy to move it across if you feel strongly about it though 😄 

> Oh I see! That's unfortunate 😞 Perhaps `variableDefs`, `variablesSet` or `variablesList`? I have nothing against `exportVariables` -- I am just wondering whether we could maintain (somewhat wonky) symmetry with `derivedRoles` because we don't call it `exportDerivedRoles`.

The difficulty I have with those is that they clash a bit with the top-level `variables` field. Any of them could equally be used to describe the "local" variable definitions, so for example if you are starting to author a policy with autocomplete in your editor, I don't think it would be obvious which to choose.

I don't know if it's too soon to introduce `api.cerbos.dev/v2`... if we did that we could move the `variables` and `importVariables` fields down into the other policy types, and call this policy type `variables`. We could add a `cerbos migrate` type command to automate the update for users.

>I don't think we should support importing other variables. That could lead to complicated chains of imports and potentially even cyclic references. To handle use cases where people might want to have import chains, maybe we should allow variable definitions to refer to other variables that don't exist within the file?

Agreed.

> Again, rather unfortunately, we would have to support a `variables` section within the variables policy itself. It shouldn't be too bad other than being just a bit odd.

What would that look like - would we inline the locally-defined `variables` into the `definitions`? Does this potentially become confusing if variables policies can also refer to variables from other files?

My feeling is that if we're not going to support `importVariables` in variables policies despite it being a top-level field, we could equally not support `variables` in variables policies.

> What do we do about duplicate variable definitions? Should we just let the last definition win or simply refuse to compile?

Duplicate derived roles is an error so I think it is reasonable for duplicate variable definitions to also be an error.

**Comment by charithe** [comment]
Fair enough, `exportVariables` would have to do then.

I am hesitant to jump to API v2 just for supporting this. Even with automated tooling, it will be quite a big thing to manage both for our existing users and ourselves. We'll have to have two code paths for handling the different versions, update examples and documentation everywhere, figure out a deprecation timeline and so on. 


> What would that look like - would we inline the locally-defined `variables` into the `definitions`? Does this potentially become confusing if variables policies can also refer to variables from other files?

I was thinking we could just inline them. Otherwise, validating that an `exportVariables` policy does not have a `variables` section would be hard to enforce through other tools. For instance, we can't impose that restriction through the JSON schemas unless we create a special case for that in the code generator 😬.

> Does this potentially become confusing if variables policies can also refer to variables from other files?

Absolutely. We'll need to come up with a good set of rules defining how the variables are shadowed.

**Comment by charithe** [comment]
> Would the same argument not apply for the `importVariables` field, though?

If we add it to the top level `Policy` message, yes. I think we should push it down to individual `oneOf` messages for derived roles, principal policies and resource policies.

**Comment by haines** [comment]
> I think it was a mistake to add `variables` to the top-level. If we design this in such a way as to support both imported and inline variables only on policy types where that makes sense, we can deprecate the `variables` field.

Yes, agreed, I would prefer this approach.

Bikeshedding... how about this?

```proto
message Variables {
  repeated string import = 1;
  map<string, string> local = 2;
}
```

However... if we only put a nested `variables` field on the policy types where it makes sense (i.e. not `exportVariables`), then I don't think we should support the top-level `variables` field for `exportVariables`. There would be no migration path for when we eventually do move to API v2 and drop the top-level field.

I think this is ok, because I don't see a lot of value in supporting variables within `exportVariables` policies, and it adds complexity if we have to process the definitions to inline any local variables.

I think if we mark the top-level field as deprecated, and document it as not supported in `exportVariables` policies, it shouldn't be too big an issue. If we add title and description fields to the JSON schemas (https://github.com/cerbos/cerbos/issues/746) then folks using the YAML language server would be able to see that documentation directly in their editor.

**Review by charithe — APPROVED** [review]
Phenomenal! 🙌🏽

File: client/model.go [hunk 1]
```diff
@@ -906,6 +922,18 @@ func (rp *ResourcePolicy) AddResourceRules(rules ...*ResourceRule) *ResourcePoli
 	return rp
 }
 
+// WithVariablesImports adds import statements for exported variables.
+func (rp *ResourcePolicy) WithVariablesImports(name ...string) *ResourcePolicy {
```
[the remark this answers is not in the document]
Comment (haines) [reply client/model.go:926]: It's definitely a bit clunky but I was going for symmetry with `WithDerivedRolesImports`. Do you think we should keep them the same, or change this one to `WithVariableImports`?

[the remark this answers is not in the document]
Comment (charithe) [reply client/model.go:926]: I see your point. It should have been `WithDerivedRoleImports` but it's too late now. Naming is hard 😞

Let's keep `WithVariablesImports` then.

**Pull request merged**